feat(gux-advanced-dropdown): add property to control dropdown max-height

exsiccator opened the pull request:
COMUI-512

daragh-king-genesys commented on src/components/stable/gux-advanced-dropdown/gux-advanced-dropdown.less:
```diff
@@ -110,9 +109,12 @@ div.gux-dropdown {
 
   .gux-dropdown-options {
+    // Maximum options without scrolling which is controlled by the size property
+    max-height: calc(var(--max-options) * 32px);
```
Assuming this 32px is based on the height of `gux-dropdown-option` would it be worth it to create a common variable instead of hardcoding it twice?
I don't think we have any instance of component level less variabl files for components and their sub components but it might be useful in situations like this. Personally I am 50-50 if it is a good idea or not so I will leave it up to you.

exsiccator replied:
Yeah, the 32px is based on the height of the `gux-dropdown-option` height.  I tried using a less variable but there were some additional complications with using the variable within the calc function (potentially solvable with string interpolation somehow but I couldn't get it to work right).  As shown here, I could use a CSS variable set at the `gux-advanced-dropdown` level.

exsiccator replied:
I just pushed up a new commit using a CSS variable for the option height as well.

daragh-king-genesys commented on src/components/stable/gux-advanced-dropdown/gux-advanced-dropdown.tsx:
```diff
@@ -55,6 +55,12 @@ export class GuxAdvancedDropdown {
+   * The max number of options to display without scrolling
+   */
+  @Prop()
+  size: number = 10;
```
This default seems big to me. Did it come from UX?

exsiccator replied:
This is based on the guidelines for dropdowns currently on the Spark site.  I didn't get specific UX input related to this ticket.

daragh-king-genesys reviewed (approved):
I like the css variable approach it is very neat.

MattCheely reviewed (changes requested):
I don't think number of elements to show is the right API for the size limit for a few reasons:

- The goal of changing a size limit is usually to keep things from overflowing the page, but doing that based on number of items requires an extra step of calculation to go from item count to something like pixels, that then has to be compared to the screen size and converted back into a count.  As a consumer of the component, IMO it would be much easier to ensure my options don't overflow if I can just directly set a CSS height. That lets me leverage things like `calc`, percentage and `vh` units, `max-height` etc.
- It makes a hard assumption all elements in a single dropdown will always be the same size.  That's probably true most of the time, but if there are exceptions that come up in the future, this mechanism just won't work.
- It makes a soft assumption that elements across dropdowns are generally the same size in the default item height it bundles.  That can be changed via css custom properties from outside the component, but in that case keeping content on screen requires keeping multiple properties in sync as a design evolves (option height and item count), where just providing a height limit via CSS is only a single configuration that is stable to changes in the size of items in the list.

In the past, I've pushed back on dropdown APIs that look like:
```html
<gux-dropdown-thingy>
  <options-container>
     <option-item>...</option-item>
     <option-item>...</option-item>
  </options-container>
</gux-dropdown-thingy>
```
With the main objection being "What is `options-container` for, from the developer's perspective?"  I think having it as a CSS target for max-height is a good enough answer that would make this approach viable in the future, but for this component, that would probably be a breaking change.

So, I'd have the component take a `max-options-height` attribute that's just a css string that will be set as the max-height on the option container for now.  A future major version release would migrate to the pattern above (or some other better idea we haven't considered yet)

exsiccator commented:
> I don't think number of elements to show is the right API for the size limit for a few reasons:
> 
> * The goal of changing a size limit is usually to keep things from overflowing the page, but doing that based on number of items requires an extra step of calculation to go from item count to something like pixels, that then has to be compared to the screen size and converted back into a count.  As a consumer of the component, IMO it would be much easier to ensure my options don't overflow if I can just directly set a CSS height. That lets me leverage things like `calc`, percentage and `vh` units, `max-height` etc.
> * It makes a hard assumption all elements in a single dropdown will always be the same size.  That's probably true most of the time, but if there are exceptions that come up in the future, this mechanism just won't work.
> * It makes a soft assumption that elements across dropdowns are generally the same size in the default item height it bundles.  That can be changed via css custom properties from outside the component, but in that case keeping content on screen requires keeping multiple properties in sync as a design evolves (option height and item count), where just providing a height limit via CSS is only a single configuration that is stable to changes in the size of items in the list.
> 
> In the past, I've pushed back on dropdown APIs that look like:
> 
> ```
> <gux-dropdown-thingy>
>   <options-container>
>      <option-item>...</option-item>
>      <option-item>...</option-item>
>   </options-container>
> </gux-dropdown-thingy>
> ```
> 
> With the main objection being "What is `options-container` for, from the developer's perspective?" I think having it as a CSS target for max-height is a good enough answer that would make this approach viable in the future, but for this component, that would probably be a breaking change.
> 
> So, I'd have the component take a `max-options-height` attribute that's just a css string that will be set as the max-height on the option container for now. A future major version release would migrate to the pattern above (or some other better idea we haven't considered yet)

In my opinion, the component should be responsible for ensuring there is no overflow without **any** configuration from the user. One way I have seen this accomplished is by setting the dropdown max-height to less than 50vh and change the direction of the popper based on the location of the element in the view port. As far as I know, we do not currently have UX guidelines for this kind of complete solution.

Configuring the dropdown height by a user should be for other reasons: aesthetics, UX, or weird use cases. I have implemented the max-height based on the guidelines given in the Spark documents (scroll after 10 items). I do think that this approach has a purpose as a secondary constraint on height but is not the complete solution.

Do you want me to pursue a more complete solution like I described? Should we involve UX?

MattCheely commented:
Default max-height based on 10 items is definitely the way to go. Good call there, and thanks for catching it from the spark docs.  I agree automatic overflow prevention is a feature we should have, but I'm not saying we need that here.

Regarding the exposed API for setting height, I think we agree that there are a number of reasons someone might need to constrain the dropdown height. It might be to prevent overflow (since we don't have it automatically yet), to show a specific item count, to avoid covering some other part of the interface, aesthetic reasons, or some other thing I haven't thought of.  What I'm saying is that having the public interface for overriding the default height be "# of items" makes some or most of those use cases more difficult, because you have to do more calculations and there's only a discrete set of heights available.  The converse isn't true though. It's not very hard to get from a desired # of items to CSS units. It might make sense to have a lignt-dom css target for overall height as well as a # of items attribute in our future dropdown work, and let one take precedence. In the meantime though, if we're only exposing one, it should just be regular CSS IMO.

I don't feel incredibly strongly about it though, if other folks also think the # of items API is preferable, I'm happy to be overruled.

daragh-king-genesys commented:
@exsiccator and @MattCheely what is happening with this PR? @exsiccator if you are too busy with feature work to continue I can take it up. Just need to know what the plan is :).

exsiccator commented:
> @exsiccator and @MattCheely what is happening with this PR? @exsiccator if you are too busy with feature work to continue I can take it up. Just need to know what the plan is :).

Feel free to take it up. I have been kind of consumed with other tasks. Otherwise I will try to return this maybe next week?

MattCheely commented on src/components/stable/gux-advanced-dropdown/gux-advanced-dropdown.tsx:
```diff
@@ -186,6 +192,7 @@ export class GuxAdvancedDropdown {
 
             <div
               class="gux-dropdown-options"
+              style={{ maxHeight: this.dropdownHeight }}
```
There's an argument to be made that this would be easier to manage in some cases as a max height for `gux-dropdown-menu-container`, but it could create some broken layouts if made too small, so I'm ok with this for this iteration of the dropdown.
